File: #39_Combination_Sum/Solution.py
class Solution:
    # @param {integer[]} candidates
    # @param {integer} target
    # @return {integer[][]}
    def combinationSum(self, candidates, target):
        if not candidates:
        	return []
        candidates.sort()
        return self.bt(candidates, target)


    def bt(self, candidates, target):
    	if not candidates:
        	return []
        result = []
        sum = 0
        c = 0
        while sum <= target:
        	init = [candidates[0] for x in range(c)]
        	if target == sum:
        		result.append(init)
        	else:    		
	        	re = self.combinationSum(candidates[1:], target - sum)
	        	if re:
	        		for list in re:
		        		result.append(init + list)
        	sum += candidates[0]
        	c += 1
        return result

sol = Solution()
print sol.combinationSum([2, 3, 6, 7], 7)

File: #39_Combination_Sum/Solution.java
import java.util.List;
import java.util.Arrays;
import java.util.LinkedList;

public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (target <= 0) return null;
        if (candidates == null || candidates.length == 0) return null;
        int[] copy = Arrays.copyOf(candidates, candidates.length);
        Arrays.sort(copy);
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        for (int i = 0; i < copy.length; i++) {
            LinkedList<Integer> match = new LinkedList<Integer>();
            backtrace(copy, target, i, res, match);
        }
        return res;
    }
    
    private void backtrace(int[] array, int target, int start, List<List<Integer>> res, LinkedList<Integer> match) {
        if (start < 0 || start >= array.length) throw new IllegalArgumentException();
        if (array[start] > target) return;
        LinkedList<Integer> local = new LinkedList<Integer>(match);
        local.add(array[start]);
        if (array[start] == target) {
            res.add(new LinkedList(local));
            local.pollLast();
            return;
        }
        for (int i = start; i < array.length; i++) {
            backtrace(array, target - array[start], i, res, local);
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] a = {2, 3, 6, 7};
        List<List<Integer>> res = sol.combinationSum(a, 7);
        for (List<Integer> list: res) {
            for (Integer i : list) {
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }
}